internal/tofu/eval_variable.go

```diff
@@ -477,3 +477,29 @@ You can correct this by removing references to sensitive values, or by carefully
 		FailureMessage: errorMessage,
 	}, diags
 }
+
+// evalVariableDeprecation checks if a variable is deprecated and if so it returns a warning diagnostic to be shown to the user
+func evalVariableDeprecation(addr addrs.AbsInputVariableInstance, config *configs.Variable, expr hcl.Expression, ctx EvalContext) tfdiags.Diagnostics {
+	if config.Deprecated == "" {
+		log.Printf("[TRACE] evalVariableDeprecation: variable %s is having no deprecation configured", addr)
+		return nil
+	}
+	// if the variable is not given in the module call, do not show a warning
+	if expr == nil {
+		log.Printf("[TRACE] evalVariableDeprecation: variable %s is marked as deprecated but is not used", addr)
+		return nil
+	}
+	val := ctx.GetVariableValue(addr)
+	if val == cty.NilVal {
```

Comment:
I wonder if this check makes sense.

```hcl
module "foo" {
  field = null
}```

still represents a value and usage (of null) for the input

Reply:
This is based on the information from the [RFC](https://github.com/opentofu/opentofu/pull/2180/files#diff-f75ee43ad75dd61d95c52383a325c31bee32a3b22df43e8a523dc6b2dba6f4d7R65).
Maybe we could change that if we really want to warn on these cases.